src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -1335,10 +1336,42 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsFinal(
         /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
         /// with level > 0 then we won't post-process this part, and if num_streams > 1 we
         /// can use parallel select on such parts.
+
         bool no_merging_final = do_not_merge_across_partitions_select_final &&
             std::distance(parts_to_merge_ranges[range_index], parts_to_merge_ranges[range_index + 1]) == 1 &&
-            parts_to_merge_ranges[range_index]->data_part->info.level > 0 &&
-            data.merging_params.is_deleted_column.empty() && !reader_settings.read_in_order;
+            parts_to_merge_ranges[range_index]->data_part->info.level > 0 && !reader_settings.read_in_order;
+
+        // if do_not_merge_across_partitions_select_final = 1 and the partition contains a single part,
+        // there is no reason to handle `is_deleted` using slow inorder reading. Faster to do it with prewhere
+        if (no_merging_final && !data.merging_params.is_deleted_column.empty())
+        {
+            auto equal_function = FunctionFactory::instance().get("equals", context);
+
+            if (!prewhere_info)
+            {
+                prewhere_info = std::make_shared<PrewhereInfo>();
+            }
+
+            prewhere_info->prewhere_column_name = data.merging_params.is_deleted_column;
+
+            if (std::find(origin_column_names.begin(), origin_column_names.end(), data.merging_params.is_deleted_column) == origin_column_names.end())
+            {
+                origin_column_names.push_back(data.merging_params.is_deleted_column);
+            }
+
+            const auto * constant_node = &prewhere_info->prewhere_actions.addColumn(
+                ColumnWithTypeAndName{
+                    ColumnUInt8::create(0),
+                    std::make_shared<DataTypeUInt8>(),
+                    ""
+                });
+
+            const auto & input = &prewhere_info->prewhere_actions.addInput(data.merging_params.is_deleted_column, std::make_shared<DataTypeUInt8>());
```

Comment:
May be it is a good idea to add a comment like
'keep records with is_deleted is false'